src/Evolution/Systems/Cce/IntegrandInputSteps.hpp

```diff
@@ -59,4 +191,112 @@ using pre_computation_tags =
                Tags::EthEthRDividedByR, Tags::EthEthbarRDividedByR,
                Tags::BondiK, Tags::OneMinusY, Tags::BondiR>;
 
+/*!
+ * \brief A type list for the set of tags computed by the set of
+ * template specializations of `ComputePreSwshDerivatives`.
+ *
+ * \details This is provided for easy and maintainable construction of a
+ * `Variables` or \ref DataBoxGroup with all of the quantities needed for a CCE
+ * computation or component. The data structures represented by these tags
+ * should each have size `number_of_radial_points *
+ * Spectral::Swsh::number_of_swsh_collocation_points(l_max)`. All of these tags
+ * (for a given integrated Bondi quantity) may be computed at once if using a
+ * \ref DataBoxGroup using the template
+ * `mutate_all_pre_swsh_derivatives_for_tag` or individually using the template
+ * specializations of `ComputePreSwshDerivatives`. The full set of integrated
+ * Bondi quantities is available from the typelist
+ * `bondi_hypersurface_step_tags`.
+ */
+template <typename Tag>
+using pre_swsh_derivative_tags_to_compute_for =
```

Comment:
To make a struct with a `type` alias, how about this:

```cpp
template <typename Tag>
using pre_swsh_derivative_tags_to_compute_for = tmpl::type_<typename detail::TagsToComputeForImpl<Tag>::pre_swsh_derivative_tags>;
```

Does this work, and is perhaps more concise that writing the `struct`?

Reply:
Gave it a try, and apparently this technique doesn't give the kind of indirection granted by defining a separate struct, so performing this type alias trick doesn't appear (at least in this context) to avoid the `tmpl::bind`s.

Reply:
Hm then I don't quite understand how it works, since this is how `tmpl::type_` is defined:
```cpp
template<typename T> struct type_ { using type = T; };
```
(https://github.com/edouarda/brigand/blob/523ee302474447c808b688607b7353ffd3e5d4bd/include/standalone/brigand.hpp#L155)

Reply:
the rough understanding I glean from the compiler errors is when in the evaluation the lambda parameter `tmpl::_1` is stuck in the desired metafunction. In the case where we are evaluating @nilsdeppe 's construction, the `tmpl::_1` goes in the `Tag` template parameter, so nothing actually obtains the `pre_swsh_derivative_tags_to_compute_for<tmpl::_1>::type` (which doesn't evaluate to a type), instead the brigand apply chain expands it to `pre_swsh_derivative_tags_to_compute_for<Cce::Tags::BondiBlah>::type` (which does evaluate). However, the subsequent suggestion (by @nilsleiffischer , to disambiguate) will cause the `typename detail::TagsToComputeForImpl<tmpl::_1>::pre_swsh_derivative_tags` to be explicitly evaluated (because it is passed as the template argument itself to `tmpl::type_`, and that type doesn't exist.
To be fair, this is such a subtle piece of this logic that I find that in most instances I have to try the thing I'm going for with combinations of bind and pin before I work out what is actually supposed to happen.

Reply:
@moxcodes is correct. A type alias is an "eager" metafunction (it gets evaluated immediately), while a `struct` with the `type` type alias does not have the call invoked until `T::type` is retrieved (it is a "lazy" metafunction). This is the cornerstone of TMP libraries and also very confusing. `bind` basically is a way of turning an eager metafunctions into lazy ones.

Now `pin` tells brigand that the template class should not be evaluated. For example, by default if you want a `tmpl::list<A>` in a tmpl call it will treat `list` as a metafunction rather than a container. To mark something as a "container" you use `pin`. Does that help?

Reply:
Yes this was very helpful :) thanks @moxcodes @nilsdeppe